src/Web/Promise/Lazy.purs

```diff
@@ -68,3 +68,12 @@ race :: forall a. Array (LazyPromise a) -> LazyPromise a
 race as = LazyPromise do
   as' <- traverse (\(LazyPromise a) -> a) as
   runEffectFn1 P.race as'
+
+fromPromise :: forall a. Effect (P.Promise a) -> LazyPromise a
+fromPromise p = LazyPromise $ runEffectFn2 P.then_ (mkEffectFn1 (pure <<< P.resolve <<< Box)) =<< p
+
+toPromise :: forall a. LazyPromise a -> Effect (P.Promise a)
+toPromise (LazyPromise p) = unbox =<< p
+  where
+  unbox :: forall b. P.Promise (Box b) -> Effect (P.Promise b)
+  unbox = runEffectFn2 P.then_ (mkEffectFn1 \(Box b) -> pure (P.resolve b))
```

Comment:
An unintuitive issue here is promise flattening. `LazyPromise` uses `Box` to add structure and prevent flattening. So removing it will reintroduce flattening.

Reply:
Does inferring the type for `toPromise` add the `Flatten` constraint?

Reply:
It doesn't infer the constraint. Are you envisioning something like this implementation?

```purs
toPromise :: forall a. Flatten a b => LazyPromise a -> Effect (P.Promise b)
toPromise (LazyPromise p) = runEffectFn2 P.then_ (mkEffectFn1 \(Box b) -> pure (P.resolve b)) =<< p
```

or 

```purs
toPromise :: forall a. Flatten a a => LazyPromise a -> Effect (P.Promise a)
```

I'm unfortunately not familiar with promise flattening 😬.

Reply:
Ah, I assumed the module `P` referred to `Web.Promise`. This is where the type-safe functions that model flattening are.

I would recommend using `Web.Promise.then_` and `Web.Promise.resolve` for these new functions, instead of the internal unsafe functions.

Reply:
Does this inferred type signature look correct to you?

```
toPromise :: forall a b c. Flatten a c => Flatten c b => LazyPromise a -> Effect (P.Promise b)
toPromise (LazyPromise p) = Promise.then_ (\(Box b) -> pure (Promise.resolve b)) =<< p
```

Reply:
The double flatten constraint looks redundant to me. It looks like the unnecessary one is coming from the `Promise.then_` constraint. How about:

```purescript
toPromise :: forall a b. Flatten a b => LazyPromise a -> Effect (P.Promise b)
toPromise (LazyPromise p) = P.then_ (\(Box b) -> pure (Promise.resolve b)) =<< p
```

Reply:
I think this is now basically identical to your version here: https://github.com/purescript-web/purescript-web-promise/pull/12/files#r655780671

Reply:
Mind taking another look at the latest?